table.go

```diff
@@ -230,7 +287,21 @@ func (rt *RoutingTable) removePeer(p peer.ID) {
 	if bucket.remove(p) {
 		// peer removed callback
 		rt.PeerRemoved(p)
-		return
+
+		// remove this bucket if it was the last bucket and it's now empty
+		// provided it isn't the ONLY bucket we have.
+		if len(rt.buckets) > 1 && bucketID == len(rt.buckets)-1 && len(bucket.peers()) == 0 {
+			rt.buckets[bucketID] = nil
+			rt.buckets = rt.buckets[:bucketID]
+			return
+		}
+
+		// if the second last bucket just became empty, remove and replace it with the last bucket.
+		if bucketID == len(rt.buckets)-2 && len(bucket.peers()) == 0 {
```

Comment:
I don't think this is good enough, as is shown in my test comment. What happens if we have buckets 1,2,3,4 and remove the peers from buckets 2 then 3 then 4? Here we'll collapse bucket 4 into bucket 3 and bucket 3 into bucket 2, leaving us with two total buckets instead of just one.

Perhaps doing this in a for loop would be better so we can close all buckets at the end that are empty.

Reply:
I agree and this has been fixed.